### PR TITLE
Compiler fixes

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/core/src/linker/view.ts [hunk 1]
```diff
@@ -351,7 +351,7 @@ export abstract class AppView<T> {
     }
   }
 
-  eventHandler(cb: Function): Function { return cb; }
+  eventHandler<E, R>(cb: (e?: E) => R): (e?: E) => R { return cb; }
```
Comment (vicb) [inline modules/@angular/core/src/linker/view.ts]: e -> event ?

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 2]
```diff
@@ -304,9 +304,13 @@ export class CompileMetadataResolver {
                 .map(type => this.getTypeMetadata(type, staticTypeModuleUrl(type))));
       }
       if (meta.bootstrap) {
-        bootstrapComponents.push(
-            ...flattenArray(meta.bootstrap)
-                .map(type => this.getTypeMetadata(type, staticTypeModuleUrl(type))));
+        bootstrapComponents.push(...flattenArray(meta.bootstrap).map(type => {
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]:
```ts
const typeMetadata = flattenArray().map(...);
bsC.push(...typeMetadata);
```

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 3]
```diff
@@ -304,9 +304,13 @@ export class CompileMetadataResolver {
...
+          if (!isValidType(type)) {
+            throw new Error(
+                `Unexpected value '${stringify(type)}' used in bootstrap of module '${stringify(moduleType)}'`);
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]: `used for the bootstrap property of '${...}'` ?

**Comment by vicb** [comment]
LGTM with nits.

Does this PR close opened GH issues ? (if yes please link them from the commit comments)

File: modules/@angular/compiler-cli/integrationtest/src/module.ts [hunk 4]
```diff
@@ -15,18 +15,33 @@ import {MultipleComponentsMyComp, NextComp} from './a/multiple_components';
...
+    CompWithNgContent,
+    CompUsingRootModuleDirectiveAndPipe,
+    CompWithProviders,
+    CompWithReferences,
+    CompUsingPipes,
+    CompConsumingEvents,
+    DirPublishingEvents,
+    MultipleComponentsMyComp,
+    DirectiveForQuery,
+    NextComp
```
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/src/module.ts]: Done

File: modules/@angular/core/src/linker/view.ts (hunk 1 above)
Comment (tbosch) [inline modules/@angular/core/src/linker/view.ts]: Done

File: modules/@angular/compiler/src/metadata_resolver.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/metadata_resolver.ts]: Done

File: modules/@angular/compiler/src/metadata_resolver.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/compiler/src/metadata_resolver.ts]: Done

**Comment by tbosch** [comment]
No open issues.

**Comment by mprobst** [comment]
Merged as ea95c391c1c8ff754610e3cb301651c9aff3693e.

**Pull request closed without merging**